### PR TITLE
Implement CloudWatch reporter

### PR DESCRIPTION
**Pull request by jkbudd**
Starport was reliant on this library for reporting metrics to CloudWatch - https://github.com/azagniotov/codahale-aggregated-metrics-cloudwatch-reporter

This PR ports that code over to Scala (except for CollectionsUtils). 

It's a fairly rough and ready port, a lot can be done to make the code more idiomatic.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -18,10 +18,10 @@ val awsSdkSNS              = "com.amazonaws"          %  "aws-java-sdk-sns"
 val awsSdkCloudWatch       = "com.amazonaws"          %  "aws-java-sdk-cloudwatch"     % awsSdkVersion
 val stubbornArtifact       = "com.krux"               %% "stubborn"             % "1.4.1"
 val metricsGraphite        = "io.dropwizard.metrics"  %  "metrics-graphite"     % "4.1.12.1"
+val metricsJvm             = "io.dropwizard.metrics"  %  "metrics-jvm"          % "4.1.12.1"
```
Comment (realstraw) [inline build.sbt:21]: is this required?

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala [hunk 2]
```diff
@@ -0,0 +1,840 @@
...
+  * <p>
+  * Use {@link CloudWatchReporter.Builder} to construct instances of this class. The {@link CloudWatchReporter.Builder}
+  * allows to configure what aggregated metrics will be reported as a single {@link MetricDatum} to CloudWatch.
+  * <p>
+  * There are a bunch of {@code with*} methods that provide a sufficient fine-grained control over what metrics
+  * should be reported
+  */
+case class CloudWatchReporter(builder: Builder)
+    extends ScheduledReporter(builder.metricRegistry,
+                              "coda-hale-metrics-cloud-watch-reporter",
```
Comment (realstraw) [inline starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:26]: can you make sure it's formated using https://github.com/krux/starport/blob/master/.scalafmt.conf

File: build.sbt (hunk 1 above)
Comment (realstraw) [reply build.sbt:21]: maybe we can remove this if it's not needed

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala (hunk 2 above)
Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala]: Apologies, now formatted correctly.

File: build.sbt (hunk 1 above)
Comment (jkbudd) [reply build.sbt:21]: That's there so we can report JVM metrics. I can remove that if we don't want to send those to CloudWatch.

File: starport-core/src/main/java/com/krux/starport/metric/CollectionsUtils.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class CollectionsUtils {
+
+    public static <T> Collection<List<T>> partition(final Collection<T> wholeCollection, final int partitionSize) {
+        final int[] itemCounter = new int[]{0};
+
+        return wholeCollection.stream()
```
Comment (jkbudd) [inline starport-core/src/main/java/com/krux/starport/metric/CollectionsUtils.java:12]: @realstraw I think this is the same as Scala's `grouped`. Have I got that right?

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala [hunk 4]
```diff
@@ -0,0 +1,925 @@
...
+  }
+
+}
+
+object CloudWatchReporter {
+
+  private val LOGGER: Logger = LoggerFactory.getLogger(classOf[CloudWatchReporter])
+
+  // Visible for testing
+  val DIMENSION_NAME_TYPE: String = "Type"
```
Comment (realstraw) [inline starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:538]: let's use `CamelCase` if we would like these to be scala

File: build.sbt (hunk 1 above)
Comment (realstraw) [reply build.sbt:21]: don't think we are reporting jvm metrics. I would recommend stay with original implementation or have a very simplified one with only metrics that we are using.

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala (hunk 4 above)
Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala]: Done.

File: build.sbt (hunk 1 above)
Comment (jkbudd) [reply build.sbt:21]: OK cool. You are correct that we are not reporting JVM metrics to CloudWatch currently.

Comment (jkbudd) [reply build.sbt]: I have removed JVM metric reporting.

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala [hunk 5]
```diff
@@ -0,0 +1,925 @@
...
+import com.codahale.metrics.jvm.{
+  BufferPoolMetricSet,
+  ClassLoadingGaugeSet,
+  FileDescriptorRatioGauge,
+  GarbageCollectorMetricSet,
+  MemoryUsageGaugeSet,
+  ThreadStatesGaugeSet
+}
+import org.slf4j.{Logger, LoggerFactory}
+import CloudWatchReporter._
```
[the remark this answers is not in the document]
Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala]: OK fixed.

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala [hunk 6]
```diff
@@ -0,0 +1,925 @@
...
+/**
+ * Reports metrics to <a href="http://aws.amazon.com/cloudwatch/">Amazon's CloudWatch</a> periodically.
+ * <p>
+ * Use {@link CloudWatchReporter.Builder} to construct instances of this class. The {@link CloudWatchReporter.Builder}
+ * allows to configure what aggregated metrics will be reported as a single {@link MetricDatum} to CloudWatch.
+ * <p>
+ * There are a bunch of {@code with*} methods that provide a sufficient fine-grained control over what metrics
+ * should be reported
+ */
+case class CloudWatchReporter(builder: Builder)
```
[the remark this answers is not in the document]
Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala]: Got it. It's now a plain old class.

File: starport-core/src/main/scala/com/krux/starport/metric/DimensionedNameBuilder.scala [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package com.krux.starport.metric
+
+import com.amazonaws.services.cloudwatch.model.Dimension
+
+import scala.collection.mutable.Map
+
+case class DimensionedNameBuilder(
```
[the remark this answers is not in the document]
Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/DimensionedNameBuilder.scala:7]: OK this is now a plain class as well.

File: starport-core/src/main/scala/com/krux/starport/metric/DimensionedNameBuilder.scala [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  name: String,
+  var dimensions: Map[String, Dimension] = Map().empty
```
[the remark this answers is not in the document]
Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/DimensionedNameBuilder.scala]: Fixed - it's a val now.

**Comment by jkbudd** [comment]
@realstraw I've addressed the comments so far. Any more feedback?

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala [hunk 9]
```diff
@@ -0,0 +1,889 @@
...
+      try cloudWatchAsyncClient.shutdown()
+      catch {
+        case e: Exception =>
+          logger.error("Error shutting down AmazonCloudWatchAsync", e)
+
+      }
+    }
+  }
+
+  private def processGauge[_ <: Numeric[_]](
```
Comment (realstraw) [inline starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:163]: Is this scala Numeric? If that's the case, you should not do type casting here with `value.asInstanceOf[Double]` but instead use `implicitly[T].toDouble(value)`

Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:163]: That's a very good catch! Thanks for that.

Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala]: @realstraw I took another look at this and rewrote it using more idiomatic Scala. Some notes: 

1. The method was trying to get a value from the gauge, which could be `null`, and so was wrapping it in an `Optional`. I used a Scala `Option` to do this now.
2. A Gauge is parameterized by type, but not necessarily a numeric type. The method now checks whether the non-null value returned by `getValue` is a `java.lang.Number` and then will do the cast from a Java `Number` to a Scala `Double`. 
3. In the `null` or non-numeric case, nothing will be reported and a warning will be logged.

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala [hunk 10]
```diff
@@ -0,0 +1,886 @@
...
+/**
+ * Reports metrics to <a href="http://aws.amazon.com/cloudwatch/">Amazon's CloudWatch</a> periodically.
+ * <p>
+ * Use {@link CloudWatchReporter.Builder} to construct instances of this class. The {@link CloudWatchReporter.Builder}
+ * allows to configure what aggregated metrics will be reported as a single {@link MetricDatum} to CloudWatch.
+ * <p>
+ * There are a bunch of {@code with*} methods that provide a sufficient fine-grained control over what metrics
+ * should be reported
+ */
+class CloudWatchReporter(builder: Builder)
```
Comment (realstraw) [inline starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:42]: let's also make the constructor private e.g. `class CloudWatchReporter private (builder: Builder)` to make it clear that builder pattern should be used

File: starport-core/src/main/java/com/krux/starport/metric/CollectionsUtils.java (hunk 3 above)
Comment (realstraw) [reply starport-core/src/main/java/com/krux/starport/metric/CollectionsUtils.java:12]: so why are we not using `grouped`?

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala [hunk 11]
```diff
@@ -0,0 +1,886 @@
...
+ * Use {@link CloudWatchReporter.Builder} to construct instances of this class. The {@link CloudWatchReporter.Builder}
+ * allows to configure what aggregated metrics will be reported as a single {@link MetricDatum} to CloudWatch.
+ * <p>
+ * There are a bunch of {@code with*} methods that provide a sufficient fine-grained control over what metrics
+ * should be reported
+ */
+class CloudWatchReporter private (builder: Builder)
+    extends ScheduledReporter(
+      builder.metricRegistry,
+      "coda-hale-metrics-cloud-watch-reporter",
```
Comment (realstraw) [inline starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:45]: let's change this to `starport-cloudwatch-reporter`

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala [hunk 12]
```diff
@@ -0,0 +1,886 @@
...
+      builder.metricFilter,
+      builder.rateUnit,
+      builder.durationUnit
+    ) {
+
+  /**
+   * We only submit the difference in counters since the last submission. This way we don't have to reset
+   * the counters within this application.
+   */
+  private val lastPolledCounts: Map[Counting, Long] = new ConcurrentHashMap()
```
Comment (realstraw) [inline starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:55]: is this how dependency that we are using implementing this?

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala [hunk 13]
```diff
@@ -0,0 +1,886 @@
...
+  private val durationUnit: StandardUnit = builder.cwDurationUnit
+
+  private val highResolution: Boolean = builder.highResolution
+
+  override def report(
+    gauges: java.util.SortedMap[String, Gauge[_]],
+    counters: java.util.SortedMap[String, Counter],
+    histograms: java.util.SortedMap[String, Histogram],
+    meters: java.util.SortedMap[String, Meter],
+    timers: java.util.SortedMap[String, Timer]
```
Comment (realstraw) [inline starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:72]: let's import those `java.util.SortedMap` as `java.util.{SortedMap => JSortedMap}`

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala [hunk 14]
```diff
@@ -0,0 +1,886 @@
...
+  ): Unit = {
+
+    if (builder.dryRun) {
```
Comment (realstraw) [inline starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:75]: let's remove the dryRun mode for simplicity

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala (hunk 11 above)
Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:45]: Yeah good call.

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala (hunk 13 above)
Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:72]: Can do.

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala (hunk 14 above)
Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala:75]: OK.

File: starport-core/src/main/java/com/krux/starport/metric/CollectionsUtils.java (hunk 3 above)
Comment (jkbudd) [reply starport-core/src/main/java/com/krux/starport/metric/CollectionsUtils.java:12]: We can. I just wanted to check my understanding of what the original Java code was doing.

File: starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala (hunk 12 above)
Comment (jkbudd) [reply starport-core/src/main/scala/com/krux/starport/metric/CloudWatchReporter.scala]: Yes.

**Comment by realstraw** [comment]
LGTM now, can you bump up the version to 7.2.0?

**Comment by jkbudd** [comment]
Can do.

**Comment by jkbudd** [comment]
@realstraw This is ready to merge now.

**Pull request merged**